File: frontend/src/railway/normalizeRailwayRows.ts
import type { RailwayRow } from "./types";

function trim(value: string | null | undefined): string | null {
  if (value == null) return null;
  const text = value.replace(/\s+/g, " ").trim();
  return text || null;
}

function sentenceCase(value: string | null | undefined): string | null {
  const text = trim(value);
  if (!text) return null;
  return text.charAt(0).toUpperCase() + text.slice(1);
}

function dedupeNote(note: string | null | undefined): string | null {
  const text = trim(note);
  if (!text) return null;
  const parts = text.split(/[.;]+/).map((p) => p.trim()).filter(Boolean);
  const seen = new Set<string>();
  const unique: string[] = [];
  for (const part of parts) {
    const key = part.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    unique.push(part);
  }
  return unique.length > 0 ? unique.join(". ") : null;
}

export function normalizeRailwayRow(row: RailwayRow): RailwayRow {
  const warnings = row.warnings.map((w) => trim(w)).filter(Boolean) as string[];
  let speedLimit = row.speedLimit;
  if (speedLimit != null && speedLimit <= 0) {
    speedLimit = null;
    warnings.push("speedLimit ignored: non-positive value");
  }

  return {
    location: sentenceCase(row.location),
    assetKind: row.assetKind,
    assetNumber: trim(row.assetNumber),
    reference: trim(row.reference),
    defect: trim(row.defect),
    speedLimit,
    note: dedupeNote(row.note),
    sourceText: trim(row.sourceText) ?? "",
    warnings,
  };
}

export function normalizeRailwayRows(rows: RailwayRow[]): RailwayRow[] {
  return rows.map(normalizeRailwayRow);
}

File: backend/app/services/railway/normalize_railway_rows.py
"""Лёгкая deterministic-нормализация после LLM — без синтеза значений."""

from __future__ import annotations

import re

from app.services.railway.types import RailwayRow

_SPACE_RE = re.compile(r"\s+")


def _trim(value: str | None) -> str | None:
    if value is None:
        return None
    text = _SPACE_RE.sub(" ", value).strip()
    return text or None


def _sentence_case(value: str | None) -> str | None:
    text = _trim(value)
    if not text:
        return None
    return text[0].upper() + text[1:]


def _dedupe_note(note: str | None) -> str | None:
    text = _trim(note)
    if not text:
        return None
    parts = [p.strip() for p in re.split(r"[.;]+", text) if p.strip()]
    seen: set[str] = set()
    unique: list[str] = []
    for part in parts:
        key = part.lower()
        if key in seen:
            continue
        seen.add(key)
        unique.append(part)
    return ". ".join(unique) if unique else None


def normalize_railway_row(row: RailwayRow) -> RailwayRow:
    warnings = [_trim(w) for w in row.warnings]
    warnings = [w for w in warnings if w]

    speed = row.speed_limit
    if speed is not None and speed <= 0:
        speed = None
        warnings.append("speedLimit ignored: non-positive value")

    return RailwayRow(
        location=_sentence_case(row.location),
        asset_kind=row.asset_kind,
        asset_number=_trim(row.asset_number),
        reference=_trim(row.reference),
        defect=_trim(row.defect),
        speed_limit=speed,
        note=_dedupe_note(row.note),
        source_text=_trim(row.source_text) or "",
        warnings=warnings,
    )


def normalize_railway_rows(rows: list[RailwayRow]) -> list[RailwayRow]:
    return [normalize_railway_row(row) for row in rows]
